ably/realtime_channel.go

```diff
@@ -203,34 +252,68 @@ func (c *RealtimeChannel) mayAttach(checkActive bool) (result, error) {
 	if checkActive {
 		switch c.state {
 		case ChannelStateAttaching:
-			return c.internalEmitter.listenResult(ChannelStateAttached, ChannelStateFailed), nil
-		case ChannelStateAttached:
+			return c.internalEmitter.listenResult(ChannelStateAttached, ChannelStateDetached, ChannelStateSuspended, ChannelStateFailed), nil // RTL4h
+		case ChannelStateAttached: // RTL4a
 			return nopResult, nil
 		}
 	}
 	return c.lockAttach(nil)
 }
 
 func (c *RealtimeChannel) lockAttach(err error) (result, error) {
+	if c.state == ChannelStateFailed { // RTL4g
+		err = nil
+	}
+
 	switch c.client.Connection.State() {
 	// RTL4b
 	case ConnectionStateInitialized,
 		ConnectionStateClosed,
 		ConnectionStateClosing,
+		ConnectionStateSuspended,
 		ConnectionStateFailed:
-		return nil, newError(80000, errAttach)
+		return nil, newError(ErrConnectionFailed, errConnAttach(c.client.Connection.State()))
 	}
 
-	c.lockSetState(ChannelStateAttaching, err)
-
-	res := c.internalEmitter.listenResult(ChannelStateAttached, ChannelStateFailed)
+	sendAttachMsg := func() (result, error) {
+		res := c.internalEmitter.listenResult(ChannelStateAttached, ChannelStateDetached, ChannelStateSuspended, ChannelStateFailed) //RTL4d
+		msg := &proto.ProtocolMessage{
+			Action:  proto.ActionAttach,
+			Channel: c.Name,
+		}
+		if c.channelOpts().Params != nil {
+			msg.Params = c.channelOpts().Params
+		}
+		if c.channelOpts().Modes != nil {
+			msg.SetModesAsFlag(c.channelOpts().Modes)
+		}
+		if c.attachResume {
+			msg.Flags.Set(proto.FlagAttachResume)
+		}
+		c.client.Connection.send(msg, nil)
+		return res, nil
+	}
 
-	msg := &proto.ProtocolMessage{
-		Action:  proto.ActionAttach,
-		Channel: c.Name,
+	if c.state == ChannelStateDetaching { //// RTL4h
+		attachRes := c.internalEmitter.listenResult(ChannelStateDetached, ChannelStateFailed)
+		return resultFunc(func(ctx context.Context) error { // runs inside goroutine, need to check for locks again before accessing state
+			detachErr := attachRes.Wait(ctx)
```

Comment:
Okay, I will check changes with respect to contexts

Reply:
Yes, if you look carefully, it will unblock that `wait`, move to next line where it sends `attach` message and returns immediately. So, we are not blocking anywhere

Reply:
But we _don't_ want the user to be able to unblock a wait in the inner parts of the library. We just want the user to be able to unblock its own call to `Detach`, but the underlying process started by such call such carry on in the background.

We made this explicit decision in https://github.com/ably/ably-go/issues/197#issuecomment-696979466.

Reply:
resolved